### PR TITLE
MP-7308: Feature AutoNumber type

### PR DESCRIPTION
**Pull request by danield-ap**

### CODE REVIEW

**Review by gjorgji-grgovski-authoritypartners — APPROVED** [review]

File: src/Nox.Types/Types/AutoNumber/AutoNumber.cs
Comment (rochar) [inline src/Nox.Types/Types/AutoNumber/AutoNumber.cs:1]: Can you please rename to DatabaseNumber

File: src/Nox.Types/Types/AutoNumber/AutoNumber.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    /// </summary>
+    /// <param name="value">Value to be parsed.</param>
+    /// <returns>New <see cref="AutoNumber"/> object.</returns>
+    public new static AutoNumber From(uint value) => new() { Value = value };
+
+    /// <summary>
+    /// Gets the current Value incremented by one.
+    /// </summary>
+    /// <returns>The current value incremented by one.</returns>
+    public uint GetNext() => ++Value;
```
Comment (rochar) [inline src/Nox.Types/Types/AutoNumber/AutoNumber.cs:24]: no logic is needed, this is kind of a marker class. This type only load the value from database nothing else

File: src/Nox.Types/Types/AutoNumber/AutoNumber.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        Value = 0;
+    }
+
+    /// <summary>
+    /// Creates a <see cref="AutoNumber"/> object with the initial value of <paramref name="value"/>.
+    /// </summary>
+    /// <param name="value">Value to be parsed.</param>
+    /// <returns>New <see cref="AutoNumber"/> object.</returns>
+    public new static AutoNumber From(uint value) => new() { Value = value };
```
Comment (rochar) [inline src/Nox.Types/Types/AutoNumber/AutoNumber.cs:18]: Autonumber cannot never be created, the From should raise an exception

Should only Load from the Database

File: src/Nox.Types.EntityFramework/Types/AutoNumber/AutonumberConverter.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿using Microsoft.EntityFrameworkCore.Storage.ValueConversion;
+
+namespace Nox.Types.EntityFramework.Types;
+
+public class AutoNumberConverter : ValueConverter<AutoNumber, uint>
+{
+    public AutoNumberConverter() : base(autonumber => autonumber.Value, currentNumber => AutoNumber.From(currentNumber)) { }
```
Comment (rochar) [inline src/Nox.Types.EntityFramework/Types/AutoNumber/AutonumberConverter.cs:7]: The convention is FromDatabase... that skips validation

File: tests/Nox.Types.Tests/EntityFrameworkTests/Configuration/CountryDbConfiguration.cs [hunk 4]
```diff
@@ -41,6 +41,7 @@ public void Configure(EntityTypeBuilder<Country> builder)
         builder.Property(e => e.PageHtml).HasConversion<HtmlConverter>();
         builder.Property(e => e.CitiesCounties).HasConversion<YamlConverter>();
         builder.Property(e => e.PhoneNumber).HasConversion<PhoneNumberConverter>();
+        builder.Property(e => e.AutoId).HasConversion<AutoNumberConverter>();
```
Comment (rochar) [inline tests/Nox.Types.Tests/EntityFrameworkTests/Configuration/CountryDbConfiguration.cs:45]: needs also to be set as AutoGenerated

File: src/Nox.Types.EntityFramework/Types/AutoNumber/AutonumberConverter.cs
Comment (rochar) [inline src/Nox.Types.EntityFramework/Types/AutoNumber/AutonumberConverter.cs:1]: In the future we will need to define some options, like increment, seed, etc.. (no need on this PR)

**Review by rochar — CHANGES_REQUESTED** [review]
looks good, we just need to simplify it

File: tests/Nox.Types.Tests/EntityFrameworkTests/Configuration/CountryDbConfiguration.cs (hunk 4 above)
Comment (danield-ap) [reply tests/Nox.Types.Tests/EntityFrameworkTests/Configuration/CountryDbConfiguration.cs:49]: Property set as `ValueGeneratedOnAdd()`.

File: src/Nox.Types.EntityFramework/Types/AutoNumber/AutonumberConverter.cs (hunk 3 above)
Comment (danield-ap) [reply src/Nox.Types.EntityFramework/Types/AutoNumber/AutonumberConverter.cs]: Updated to use `FromDatabase`.

File: src/Nox.Types/Types/AutoNumber/AutoNumber.cs (hunk 2 above)
Comment (danield-ap) [reply src/Nox.Types/Types/AutoNumber/AutoNumber.cs]: `From` method now raises an exception.

File: src/Nox.Types/Types/AutoNumber/AutoNumber.cs (hunk 1 above)
Comment (danield-ap) [reply src/Nox.Types/Types/AutoNumber/AutoNumber.cs]: `GetNext` method removed. DatabaseNumber can still be accessed through `Value`.

File: src/Nox.Types/Types/AutoNumber/AutoNumber.cs
Comment (danield-ap) [reply src/Nox.Types/Types/AutoNumber/AutoNumber.cs:1]: Class renamed to DatabaseNumber. Also renamed nox type to DatabaseNumber.

**Review by rochar — APPROVED** [review]

**Pull request merged**